named_arrays/_functions/functions.py

```diff
@@ -653,11 +788,28 @@ def from_scalar_array(
 
     @property
     def axes(self) -> tuple[str, ...]:
-        return tuple(self.shape.keys())
+        return tuple(self.inputs.shape | self.outputs.shape)
 
     @property
     def shape(self) -> dict[str, int]:
-        return na.shape_broadcasted(self.inputs, self.outputs)
+        outputs_shape = self.outputs.shape
+        inputs_shape = self.inputs.shape
+
+        shape = {}
+        for axis in self.axes:
+            if axis in self.axes_vertex:
+                shape[axis] = outputs_shape[axis]
+
+            for axis in self.axes_center:
+                if axis in inputs_shape:
+                    if axis in outputs_shape:
+                        shape[axis] = max([outputs_shape[axis], inputs_shape[axis]])
+                    else:
+                        shape[axis] = inputs_shape[axis]
+                elif axis in outputs_shape:
+                    shape[axis] = outputs_shape[axis]
+
+        return shape
```

Comment:
```suggestion
@property
    def shape(self) -> dict[str, int]:
    
        axes_center = self.axes_center
        axes_vertex = self.axes_vertex
    
        outputs_shape = self.outputs.shape
        inputs_shape = self.inputs.shape
        
        outputs_shape_center = {ax: outputs_shape[ax] for ax in outputs_shape if ax in axes_center}
        inputs_shape_center = {ax: inputs_shape[ax] for ax in inputs_shape if ax in axes_center}
        
        shape_center = na.broadcast_shapes(outputs_shape_center, inputs_shape_center)
        
        outputs_shape_vertex = {ax: outputs_shape[ax] for ax in outputs_shape if ax in axes_vertex}
        
        shape = na.broadcast_shapes(
            outputs_shape_center,
            outputs_shape_vertex,
            inputs_shape_center,
        )

        return shape
```
Does this do the same thing and is it any easier to understand?